riscv/src/riscv_interpreter.rs

```diff
@@ -37,171 +37,468 @@ fn getinst(op: &RvdOperation, i1: u32, i2: u32) -> (String, i32) {
 
 /// Interprets a buffer of 32-bits RICSV instructions into a vector of decoded RISCV instructions
 /// split by field
-pub fn riscv_interpreter(code: &[u32]) -> Vec<RiscvInstruction> {
+pub fn riscv_interpreter(code: &[u16]) -> Vec<RiscvInstruction> {
     let mut insts = Vec::<RiscvInstruction>::new();
 
     // Build an RVD data tree
     let mut rvd = Rvd::new();
     rvd.init();
 
-    // For every 32-bit instruction in the input code buffer
+    // code_len is the length of the input code buffer,
+    // which can contain both 16-bit and 32-bit instructions
     let code_len = code.len();
-    for (s, inst_ref) in code.iter().enumerate().take(code_len) {
-        //println!("riscv_interpreter() s={}", s);
+
+    // code_index is the index in the code buffer, from 0 to code_len - 1
+    let mut code_index: usize = 0;
+
+    // For every 16-bit instruction in the input code buffer
+    while code_index < code_len {
+        //println!("riscv_interpreter() code_index={}", code_index);
 
         // Get the RISCV instruction
-        let inst = *inst_ref;
+        let inst = code[code_index];
+        code_index += 1;
 
-        // Ignore instructions that are zero
+        // Manage instructions that are zero
+        // As per spec, they can only be 32 bits nop instructions
+        // In case of 16 zero bits, they are used by some compilers (e.g. Go Lang compiler) to halt
+        // the system with an error
         if inst == 0 {
             // println!("riscv_interpreter() found inst=0 at position s={} (index in u32 array)", s);
-            insts.push(RiscvInstruction::nop(inst));
+            if code_index == code_len {
+                // This is the last 16 bits in the code buffer, so this must be a 16-bits invalid
+                // instruction, so we must HALT
+                insts.push(RiscvInstruction::halt(0));
+                break;
+            }
+            assert!(code_index < code_len); // There must be at least another 16 bits
```

Comment:
This should always be true, because of the previous if statement breaking

Reply:
I deleted this line